### PR TITLE
Shift chip

### PR DESCRIPTION
**Pull request by maxgillett**

### CODE REVIEW

File: alu_u32/src/shift/stark.rs [hunk 1]
```diff
@@ -14,10 +14,41 @@ where
...
+
+        let bit_base = [1, 2, 4, 8, 16, 32, 64, 128].map(AB::Expr::from_canonical_u32);
+        let pow_base = [1 << 1, 1 << 2, 1 << 4, 1 << 8, 1 << 16].map(AB::Expr::from_canonical_u32);
+        let byte_base = [1 << 24, 1 << 16, 1 << 8, 1].map(AB::Expr::from_canonical_u32);
+
+        // Check that input byte decomposition is correct
+        let byte_2: AB::Expr = local
+            .bits_2
+            .into_iter()
+            .zip(bit_base.iter().cloned())
```
Comment (dlubarov) [inline alu_u32/src/shift/stark.rs:27]: nit: could move? `zip(bit_base)`

File: alu_u32/src/shift/stark.rs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            .map(|(bit, base)| bit * base)
+            .sum();
+        builder.assert_eq(local.input_2[3], byte_2.clone());
+
+        for bit in local.bits_2.iter() {
+            builder.assert_bool(*bit);
+        }
+
+        // Check that the power of two is correct (limited to 2^31)
+        let temp_1 = (local.bits_2[0] * pow_base[0].clone())
```
Comment (maxgillett) [inline alu_u32/src/shift/stark.rs:37]: Todo before merging: make this a local variable

**Review by dlubarov — APPROVED** [review]
LGTM besides the `temp_1` column you mentioned

File: alu_u32/src/shift/stark.rs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+use p3_matrix::MatrixRowSlices;
+
+impl<F, AB> Air<AB> for Shift32Chip
+where
+    F: AbstractField,
+    AB: AirBuilder<F = F>,
+{
+    fn eval(&self, builder: &mut AB) {
+        let main = builder.main();
+        let local: &Shift32Cols<AB::Var> = main.row_slice(0).borrow();
```
Comment (vivekvpandya) [inline alu_u32/src/shift/stark.rs:16]: why outputs from Shift32Cols never constrained here?

Comment (dlubarov) [reply alu_u32/src/shift/stark.rs:16]: I think `output` should be implicitly constrained by the bus messages; i.e. this chip essentially calls out to either the mul or div chip to get the result

Comment (vivekvpandya) [reply alu_u32/src/shift/stark.rs:16]: I tried a small tests and constraints fail https://github.com/vivekvpandya/valida/commit/21a90fdda80d4fb969f82dd3f9a97642fb5ef366

**Pull request merged**